shim/shim.c

```diff
@@ -0,0 +1,621 @@
+		}
+		offset += (size_t)ret;
+	}
+	free(proxy_ctl_msg);
+}
+
+/*!
+ * Read signals received and send message in the hyperstart protocol
+ * format to outfd
+ *
+ * \param container_id Container id
+ * \param outfd File descriptor to send the message to
+ */
+void
+handle_signals(char *container_id, int outfd) {
+	int            sig;
+	char          *buf;
+	int            ret;
+	int            cmd_type;
+	struct winsize ws;
+
+	if (! container_id || outfd < 0) {
+		return;
+	}
+
+	while (read(signal_pipe_fd[0], &sig, sizeof(sig)) != -1) {
+		printf("Handling signal : %d on fd %d\n", sig, signal_pipe_fd[0]);
+		if (sig == SIGWINCH ) {
+			cmd_type = WINSIZE;
+			if (ioctl(STDIN_FILENO, TIOCGWINSZ, &ws) == -1) {
+				shim_warning("Error getting the current window size: %s\n",
+					strerror(errno));
+				continue;
+			}
+			ret = asprintf(&buf, "{\"container_id\":\"%s\", \"row\":\"%d\", \"col\":\"%d\"}",
+                                                        container_id, ws.ws_row, ws.ws_col);
+			shim_debug("handled SIGWINCH for container %s (row=%d, col=%d)\n",
+				container_id, ws.ws_row, ws.ws_col);
+		} else {
+			cmd_type = KILLCONTAINER;
+			ret = asprintf(&buf, "{\"container_id\":\"%s\", \"signal\":\"%d\"}",
+                                                        container_id, sig);
+			shim_debug("Killed container %s with signal %d\n", container_id, sig);
+		}
+		if (ret == -1) {
+			abort();
+		}
+
+		send_proxy_hyper_message(outfd, cmd_type, buf);
+		free(buf);
+        }
+}
+
+/*!
+ * Read data from infd and write to outfd
+ *
+ * \param infd File descriptor to read from
+ * \param outfd File descriptor to write to
+ */
+void
+handle_stdin(struct cc_shim *shim)
+{
+	ssize_t    nread;
+	char       buf[BUFSIZ-12];
+	char       wbuf[BUFSIZ];
+	int        ret;
+	ssize_t     len;
+
+	if (! shim || shim->proxy_io_fd < 0) {
+		return;
+	}
+
+	// write data to I/O fd
+	while ((nread = read(STDIN_FILENO, buf, BUFSIZ-12)) != -1) {
+		set_big_endian_64 ((uint8_t*)wbuf, shim->io_seq_no);
+		len = nread + STREAM_HEADER_SIZE;
+		set_big_endian_32 ((uint8_t*)wbuf, (uint32_t)len);
+		strncpy(wbuf, buf, (size_t)nread);
+
+		// TODO: handle write in the poll loop to account for write blocking
+		ret = (int)write(shim->proxy_io_fd, wbuf, (size_t)nread);
+		if (ret == -1) {
+			shim_warning("Error writing from fd %d to fd %d: %s\n",
+				STDIN_FILENO, shim->proxy_io_fd, strerror(errno));
+			return;
+		}
+	}
+}
+
+/*!
+ * Read and parse I/O message on proxy I/O fd
+ *
+ * \param shim \ref cc_shim
+ * \param[out] seq Seqence number of the I/O stream
+ * \param[out] stream_len Length of the data received
+ *
+ * \return newly allocated string on success, else \c NULL.
+ */
+char*
+read_IO_message(struct cc_shim *shim, uint64_t *seq, ssize_t *stream_len) {
+	char *buf = NULL;
+	ssize_t need_read = STREAM_HEADER_SIZE;
+	ssize_t bytes_read = 0, want, ret;
+
+	if (!shim) {
+		return NULL;
+	}
+
+	*stream_len = 0;
+
+	buf = calloc(STREAM_HEADER_SIZE, 1);
```

Comment:
No need to heap allocate the header, could just be a static variable in this function or use alloca()